sdk/identity/identity/src/credentials/workloadIdentityCredential.ts

```diff
@@ -59,7 +60,7 @@ export class WorkloadIdentityCredential implements TokenCredential {
     this.federatedTokenFilePath =
       workloadIdentityCredentialOptions.tokenFilePath || process.env.AZURE_FEDERATED_TOKEN_FILE;
     if (tenantId) {
-      checkTenantId(logger, tenantId);
+      checkTenantId(logger, tenantId); // todo: how can we avoid constructors throwing?
```

Comment:
are constructors throwing a bad thing if their inputs are invalid? Maybe I'm missing a correctness thing here. I know returning an object causes weird behavior but I thought throwing was okay?

Reply:
It's probably ok for them to throw if the input is invalid - the problem for us is that DefaultAzureCredential assumes all the constructors will create themselves fine and that all errors will be thrown from `getToken`.

So what happens here, if a credential in the chain throws on creation, the whole thing blows up. 

The way I think about it is we could do one of two things:

1. Make any constructors that are part of the DefaultAzureCredential chain as safe as possible
2. Make the DefaultAzureCredential constructor as safe as possible with a try...catch for each constructor in the chain

I think (2) is better, since (1) requires the developer knowing that credentials added to the chain should not throw in the constructor. 

What I can do then is use the `UnavailableDefaultCredential` thingy to store why it failed to create the credential. Later, it will be logged correctly in `getToken`

Reply:
So I have a question -  this will be analogous to what you noticed was happening before in our previous design right? - that the constructor was being created even though it shouldn't have been.
I think it's fine. 

What are the downsides to throwing the CredentialUnavailableError from here and making it test the next credential in the chain?

Reply:
Bcz i think that's what Java does.

Reply:
I think up until now if any credential in the chain threw from the constructor it would propagate to the DefaultAzureCredential constructor throwing but correct me if I'm wrong. AFAIK only during `getToken` will CredentialUnavailableError be handled and the next credential in the chain tried - reading #25024 those seemed to be problems we wanted to address right?

```
1. The CredentialUnavailableError or any other error thrown from the WorkloadIdentity Credential - should immediately try the next credential in the chain and not just exit after throwing.
2. The WorkloadIdentity should accept the env vars when passed in through the DefaultAzureCredential client.
3. The constructor for WorkloadIdentity should match with the constructor for DefaultAzureCredential. This was tricky to resolve as well.
4. Next, the CredentialUnavailableError should be thrown from the getToken() call and not from within the constructor, otherwise the next credential in the chain will not run.
```

Now, (2) and (3) are no longer concerns - you can have whatever constructor you want for the underlying credential, the factory function is responsible for checking the right values and calling the constructor with the right arguments. 

(1) and (4) are addressed as well

Does that help clarify at all?